cpp/src/utilities/type_checks.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/dictionary/dictionary_column_view.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <algorithm>
+
+namespace cudf {
+
+bool column_types_equal(column_view const& lhs, column_view const& rhs);
+
+struct columns_equal_fn {
+  template <typename T, typename... Arg>
+  bool operator()(Arg&&...)
+  {
+    return true;
+  }
+
+  template <typename T, CUDF_ENABLE_IF(is_dictionary<T>())>
+  bool operator()(column_view lhs, column_view rhs)
+  {
+    auto const kidx = dictionary_column_view::keys_column_index;
+    return lhs.num_children() > 0 and rhs.num_children() > 0
+             ? lhs.child(kidx).type() == rhs.child(kidx).type()
+             : true;
```

Comment:
I think this would return `true` if either `lhs` or `rhs` were empty but the other was not. 

```suggestion
    return lhs.num_children() > 0 and rhs.num_children() > 0
             ? lhs.child(kidx).type() == rhs.child(kidx).type()
             : lhs.is_empty() == rhs.is_empty();
```

I see there is a test for this case, so was that intended?

Reply:
Thanks for pointing this out. From arrow:
```python
In [16]: empty = pa.array(pd.Series([], dtype="category"), from_pandas=True)

In [18]: arr = pa.array(pd.Series([1, 1, 2], dtype="category"), from_pandas=True)

In [19]: empty.type == empty.type
Out[19]: True

In [20]: arr.type == empty.type
Out[20]: False
```
So what you suggested above is correct.